message.js

```diff
@@ -0,0 +1,264 @@
+    } else { // legacy non-strict message header
+        // name~4 type:1 id:4
+        length += 9;
+    }
+
+    // body
+    result = this.body.poolByteLength(result, message.body);
+    if (result.err) {
+        return result;
+    }
+    length += result.length;
+
+    return result.reset(null, length);
+};
+
+MessageRW.prototype.poolWriteInto = function poolWriteInto(result, message, buffer, offset) {
+    if (message.version > 0) {
+        result = this.poolStrictWriteInto(result, message, buffer, offset);
+    } else {
+        result = this.poolLegacyWriteInto(result, message, buffer, offset);
+    }
+    if (result.err) {
+        return result;
+    }
+    offset = result.offset;
+
+    // write body
+    return this.body.poolWriteInto(result, message.body, buffer, offset);
+};
+
+MessageRW.prototype.poolStrictWriteInto = function poolStrictWriteInto(result, message, buffer, offset) {
+    // version:2 type:2 name~4 id:4
+
+    // version:2 (with MSB set)
+    buffer.writeUInt16BE(message.version | 0x8000, offset);
+    offset += 2;
+
+    // type:2
+    var type = types[message.type];
+    if (type == null) {
+        return result.reset(new errors.InvalidMessageEnvelopeTypeName({
+            name: message.type
+        }));
+    }
+    buffer.writeUInt16BE(type, offset);
+    offset += 2;
+
+    // name.length:4
+    buffer.writeUInt32BE(message.name.length, offset);
+    offset += 4;
+
+    // name:name.length
+    buffer.write(message.name, offset, 'ascii');
+    offset += message.name.length;
+
+    // id:4
+    buffer.writeUInt32BE(message.id, offset);
+    offset += 4;
+
+    return result.reset(null, offset);
+};
+
+MessageRW.prototype.poolLegacyWriteInto = function poolLegacyWriteInto(result, message, buffer, offset) {
+    // name~4 type:1 id:4
+
+    // name.length:4
+    buffer.writeUInt32BE(message.name.length, offset, true);
+    offset += 4;
+
+    // name:name.length
+    buffer.write(message.name, offset, 'ascii');
+    offset += message.name.length;
+
+    // type:1
+    var type = types[message.type];
+    if (type == null) {
+        return result.reset(new errors.InvalidMessageEnvelopeTypeName({
+            name: message.type
+        }));
+    }
+    buffer.writeUInt8(type, offset);
+    offset += 1;
+
+    // id:4
+    buffer.writeUInt32BE(message.id, offset, true);
+    offset += 4;
+
+    return result.reset(null, offset);
+};
+
+MessageRW.prototype.poolReadFrom = function poolReadFrom(result, buffer, offset) {
+    var msb = buffer.readInt8(offset, true);
+    if (msb < 0) {
+        result = this.poolStrictReadFrom(result, buffer, offset);
+    } else {
+        result = this.poolLegacyReadFrom(result, buffer, offset);
+    }
+    if (result.err) {
+        return result;
+    }
+    var message = result.value;
+    offset = result.offset;
+
+    if (message.type === 'EXCEPTION') {
+        // TODO parse body as exception struct
+        return result.reset(new Error('Thrift exception'));
```

Comment:
Until you do this, you can also just parse the body into a TStruct and include that in the error so some information is available about the failure.